src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,218 +17,419 @@
+            NameAndTypePair column{ctx.column.name + ".null", std::make_shared<DataTypeUInt8>()};
+            node = std::make_shared<ColumnNode>(column, ctx.column_source);
+        },
+    },
+    {
+        {TypeIndex::Nullable, "isNotNull"},
+        [](QueryTreeNodePtr &, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `isNotNull(nullable_argument)` with `not(nullable_argument.null)`
+            NameAndTypePair column{ctx.column.name + ".null", std::make_shared<DataTypeUInt8>()};
+            auto & function_arguments_nodes = function_node.getArguments().getNodes();
+
+            function_arguments_nodes = {std::make_shared<ColumnNode>(column, ctx.column_source)};
+            resolveOrdinaryFunctionNodeByName(function_node, "not", ctx.context);
+        },
+    },
+    {
+        {TypeIndex::Tuple, "tupleElement"}, optimizeTupleOrVariantElement<DataTypeTuple>,
+    },
+    {
+        {TypeIndex::Variant, "variantElement"}, optimizeTupleOrVariantElement<DataTypeVariant>,
+    },
+};
+
+std::tuple<FunctionNode *, ColumnNode *, TableNode *> getTypedNodesForOptimization(const QueryTreeNodePtr & node)
+{
+    auto * function_node = node->as<FunctionNode>();
+    if (!function_node)
+        return {};
+
+    auto & function_arguments_nodes = function_node->getArguments().getNodes();
+    if (function_arguments_nodes.empty() || function_arguments_nodes.size() > 2)
+        return {};
+
+    auto * first_argument_column_node = function_arguments_nodes.front()->as<ColumnNode>();
+    if (!first_argument_column_node || first_argument_column_node->getColumnName() == "__grouping_set")
+        return {};
+
+    auto column_source = first_argument_column_node->getColumnSource();
+    auto * table_node = column_source->as<TableNode>();
+    if (!table_node)
+        return {};
+
+    const auto & storage = table_node->getStorage();
+    const auto & storage_snapshot = table_node->getStorageSnapshot();
+    auto column = first_argument_column_node->getColumn();
+
+    if (!storage->supportsOptimizationToSubcolumns() || storage->isVirtualColumn(column.name, storage_snapshot->metadata))
+        return {};
+
+    auto column_in_table = storage_snapshot->tryGetColumn(GetColumnsOptions::All, column.name);
+    if (!column_in_table || !column_in_table->type->equals(*column.type))
+        return {};
+
+    return std::make_tuple(function_node, first_argument_column_node, table_node);
+}
+
+/// First pass collects info about identifiers to determine which identifiers are allowed to optimize.
+class FunctionToSubcolumnsVisitorFirstPass : public InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>
 {
 public:
-    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitor>;
+    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>;
     using Base::Base;
 
-    void enterImpl(QueryTreeNodePtr & node) const
+    struct Data
```

Comment:
No need to introduce a separate class. You can store this data as fields of `FunctionToSubcolumnsVisitorFirstPass`.

Reply:
Makes sense